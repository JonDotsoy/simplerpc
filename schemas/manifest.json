{
  "type": "object",
  "properties": {
    "services": {
      "type": "array",
      "items": {
        "type": "object",
        "properties": {
          "name": {
            "type": "string"
          },
          "input": {
            "anyOf": [
              {
                "anyOf": [
                  {
                    "type": "object",
                    "properties": {
                      "$object": {
                        "type": "object",
                        "additionalProperties": {
                          "$ref": "#/definitions/manifest/properties/services/items/properties/input/anyOf/0"
                        }
                      }
                    },
                    "required": ["$object"],
                    "additionalProperties": false
                  },
                  {
                    "type": "object",
                    "properties": {
                      "$string": {
                        "type": "boolean",
                        "const": true
                      }
                    },
                    "required": ["$string"],
                    "additionalProperties": false
                  },
                  {
                    "type": "object",
                    "properties": {
                      "$number": {
                        "type": "boolean",
                        "const": true
                      }
                    },
                    "required": ["$number"],
                    "additionalProperties": false
                  },
                  {
                    "type": "object",
                    "properties": {
                      "$boolean": {
                        "type": "boolean",
                        "const": true
                      }
                    },
                    "required": ["$boolean"],
                    "additionalProperties": false
                  },
                  {
                    "type": "object",
                    "properties": {
                      "$array": {
                        "$ref": "#/definitions/manifest/properties/services/items/properties/input/anyOf/0"
                      }
                    },
                    "required": ["$array"],
                    "additionalProperties": false
                  },
                  {
                    "type": "object",
                    "properties": {
                      "$optional": {
                        "$ref": "#/definitions/manifest/properties/services/items/properties/input/anyOf/0"
                      }
                    },
                    "required": ["$optional"],
                    "additionalProperties": false
                  }
                ]
              },
              {}
            ]
          },
          "output": {
            "$ref": "#/definitions/manifest/properties/services/items/properties/input"
          }
        },
        "required": ["name"],
        "additionalProperties": false
      }
    }
  },
  "required": ["services"],
  "additionalProperties": false,
  "title": "manifest",
  "$schema": "http://json-schema.org/draft-07/schema#"
}
