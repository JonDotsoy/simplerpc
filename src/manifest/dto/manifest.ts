import type z from "zod";
import type { Manifest } from "../schemas/manifest.js";

export type ManifestDTO = z.infer<typeof Manifest>;
