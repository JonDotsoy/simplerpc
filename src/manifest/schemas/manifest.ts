import {
  array,
  lazy,
  object,
  optional,
  record,
  string,
  union,
  instanceof as instanceof_,
  ZodType,
  z,
  literal,
} from "zod";

const requireTransform = new WeakSet();

namespace ZodJsonSchema {
  type ZodJsonObjectSchema = { $object: Record<string, ZodJsonSchema> };
  type ZodJsonStringSchema = { $string: true };
  type ZodJsonNumberSchema = { $number: true };
  type ZodJsonBooleanSchema = { $boolean: true };
  type ZodJsonOptionalSchema = { $optional: ZodJsonSchema };
  type ZodJsonArraySchema = { $array: ZodJsonSchema };
  type ZodJsonSchema =
    | ZodJsonObjectSchema
    | ZodJsonStringSchema
    | ZodJsonNumberSchema
    | ZodJsonBooleanSchema
    | ZodJsonOptionalSchema
    | ZodJsonArraySchema;

  export const ZodJsonStringSchema: ZodType<any> = object({
    $string: literal(true),
  });

  export const ZodJsonNumberSchema: ZodType<any> = object({
    $number: literal(true),
  });

  export const ZodJsonBooleanSchema: ZodType<any> = object({
    $boolean: literal(true),
  });

  export const ZodJsonObjectSchema: ZodType<any> = object({
    $object: record(
      string(),
      lazy(() => ZodJsonSchema),
    ),
  });

  export const ZodJsonArraySchema: ZodType<any> = object({
    $array: lazy(() => ZodJsonSchema),
  });

  export const ZodJsonOptionalSchema: ZodType<any> = object({
    $optional: lazy(() => ZodJsonSchema),
  });

  export const ZodJsonSchema: ZodType<ZodJsonSchema> = union([
    ZodJsonObjectSchema,
    ZodJsonStringSchema,
    ZodJsonNumberSchema,
    ZodJsonBooleanSchema,
    ZodJsonArraySchema,
    ZodJsonOptionalSchema,
  ]).transform((e) => {
    requireTransform.add(e);
    return e;
  });

  export const isZodJsonSchema = (value: unknown): value is ZodJsonSchema => {
    return (
      typeof value === "object" && value !== null && requireTransform.has(value)
    );
  };
}

const Schema = union([ZodJsonSchema.ZodJsonSchema, instanceof_(ZodType)]);

export const toZodObject = (value: z.infer<typeof Schema>) => {
  if (!ZodJsonSchema.isZodJsonSchema(value)) return value;

  const parseZodObject = (
    value: z.infer<typeof ZodJsonSchema.ZodJsonSchema>,
  ): z.ZodSchema => {
    if ("$object" in value) {
      return z.object(
        Object.fromEntries(
          Array.from(Object.entries(value.$object), ([key, field]) => [
            key,
            parseZodObject(field),
          ]),
        ),
      );
    }
    if ("$string" in value) return z.string();
    if ("$number" in value) return z.number();
    if ("$boolean" in value) return z.boolean();
    if ("$optional" in value)
      return z.optional(parseZodObject(value.$optional));
    if ("$array" in value) return z.array(parseZodObject(value.$array));

    throw new Error("Cannot parse zod schema");
  };

  return parseZodObject(value);
};

export const Service = object({
  name: string(),
  input: optional(Schema),
  output: optional(Schema),
});

export const Manifest = object({
  services: array(Service),
});

export default Manifest;
