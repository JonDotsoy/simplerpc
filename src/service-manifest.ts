import { type ZodObject, z } from "zod";
import { zodToTs, printNode, createTypeAlias, withGetType } from "zod-to-ts";
import { visit, get } from "@jondotsoy/utils-js";
import type { ManifestDTO } from "./manifest/dto/manifest";
import { Manifest, toZodObject } from "./manifest/schemas/manifest";

export const createServiceManifest = (incomingManifest: ManifestDTO) => {
  const manifestChecked = Manifest.parse(incomingManifest);

  const manifest = {
    services: [
      ...Array.from(manifestChecked.services, (services) => {
        return {
          ...services,
          input: services.input ? toZodObject(services.input) : undefined,
          output: services.output ? toZodObject(services.output) : undefined,
        };
      }),
    ],
  };

  const createDeclaration = () => {
    let nodes: string[] = [];
    let count = 0;
    let nodesAlreadyMapped = new WeakMap<
      WeakKey,
      { index: number; aliasName: string }
    >();

    for (const service of visit<any, ZodObject<any>>(
      manifest.services,
      (service) => get.string(service, "_def", "typeName") === "ZodObject",
    )) {
      if (!nodesAlreadyMapped.has(service)) {
        const index = count++;
        const aliasName = `__type_${index}`;
        nodesAlreadyMapped.set(service, { index, aliasName });

        const node = zodToTs(service, aliasName).node;
        const typeAlias = createTypeAlias(node, aliasName);

        withGetType(service, (ts) => ts.factory.createIdentifier(aliasName));

        nodes.push(printNode(typeAlias));
      }
    }

    nodes.push("");

    manifest.services.forEach((e) => {
      const fn = z.function(
        z.tuple([e.input ?? z.void()]),
        z.promise(e.output ?? z.void()),
      );

      const node = zodToTs(fn, e.name).node;
      const typeAlias = createTypeAlias(node, e.name);

      nodes.push(`export ${printNode(typeAlias)}`);
    });

    return nodes.join("\n");
  };

  return {
    createDeclaration,
  };
};
