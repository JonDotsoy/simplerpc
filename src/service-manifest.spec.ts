import { expect, test } from "bun:test";
import { createServiceManifest } from "./service-manifest.js";
import { z, ZodAny, ZodSchema } from "zod";
import { Manifest, toZodObject } from "./manifest/schemas/manifest.js";
import type { ManifestDTO } from "./manifest/dto/manifest.js";

test("should parse json zod on zod object", async () => {
  const manifest = Manifest.parse({
    services: [
      {
        name: "hola",
        input: {
          $object: {
            hola: {
              $string: true,
            },
            a: {
              $array: {
                $number: true,
              },
            },
            b: {
              $optional: {
                $boolean: true,
              },
            },
          },
        },
      },
    ],
  } satisfies ManifestDTO);

  expect(toZodObject(manifest.services[0].input!)).toBeInstanceOf(ZodSchema);
});

test.only("should make a definition to export resources", async () => {
  const profile = z
    .object({
      name: z.string(),
      age: z.union([z.number(), z.string().endsWith("age")]),
    })
    .describe("profile");

  const manifest = createServiceManifest({
    services: [
      {
        name: "read",
        input: z.object({
          per: profile,
        }),
        output: z.object({}),
      },
      {
        name: "list",
        output: z
          .object({
            names: z.array(z.string()),
            profiles: z.array(profile),
          })
          .describe("abc"),
      },
      { name: "echo" },
      {
        name: "rec",
        input: {
          $object: {
            foo: {
              $string: true,
            },
          },
        },
      },
    ],
  });

  expect(manifest.createDeclaration()).toMatchSnapshot();
});
