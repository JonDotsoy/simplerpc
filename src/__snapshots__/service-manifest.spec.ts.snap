// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`should make a definition to export resources 1`] = `
"type __type_0 = {
    /** profile */
    per: {
        name: string;
        age: number | string;
    };
};
type __type_1 = {};
type __type_2 = {
    names: string[];
    profiles: {
        name: string;
        age: number | string;
    }[];
};
type __type_3 = {
    foo: string;
};

export type read = (args_0: __type_0, ...args_1: unknown[]) => Promise<__type_1>;
export type list = (args_0: void | undefined, ...args_1: unknown[]) => Promise<__type_2>;
export type echo = (args_0: void | undefined, ...args_1: unknown[]) => Promise<void | undefined>;
export type rec = (args_0: __type_3, ...args_1: unknown[]) => Promise<void | undefined>;"
`;
